lib/NormalModuleFactory.js

```diff
@@ -227,23 +230,63 @@ class NormalModuleFactory extends ModuleFactory {
 			}
 		);
 		this.hooks.resolve.tapAsync(
-			/** @type {TODO} */ ({
+			{
 				name: "NormalModuleFactory",
 				stage: 100
-			}),
+			},
 			(data, callback) => {
+				let { resolveOptions } = data;
 				const {
 					contextInfo,
 					context,
+					dependencies,
 					request,
-					resolveOptions,
 					fileDependencies,
 					missingDependencies,
 					contextDependencies
 				} = data;
-
+				const category =
+					dependencies.length > 0 ? dependencies[0].category : undefined;
 				const loaderResolver = this.getResolver("loader");
-				const normalResolver = this.getResolver("normal", resolveOptions);
+
+				if (category === Dependency.Categories.CommonJS) {
```

Comment:
Instead of a hard-coded setup here, apply `resolveOptions.byDependency` depending on `category`. To merge these objects use `cachedCleverMerge`. That's important for caching and using `"..."`.

Reply:
Ah thanks